### DISCUSSION TITLE
Database Schema

### DISCUSSION THREAD

**Discussion by matty0ung** [Brewtarget/brewtarget#554] [Q&A] [answered]
This may be a stupid question, but can anyone tell me what the bt_xxx tables are for?

**Comment by mikfire** [comment, marked as answer]
Sure. They allow us to find our data when we want to update an ingredient or something. As an example, we ship 14 water profiles by default right now. Let's say the user adds 6 profiles, and then we update our database and add a new profile. The entry in bt_water will be (id,water_id) if (15,21). On the next update if we have to fix that last water profile, we can easily find it in the user's database. 

They were introduced in 9838ed25. If you dig into `Database::updateDatabase`, there is a description of the mechanism but not much else.

I thought we had some basic docs on how to use them, as I need to remember how this works every time I look at it. Apparently, we don't. I should probably fix that.

**Comment by matty0ung** [comment]
Ah, I see, yes that makes sense now.  It's like an extra optional ID for the items of data that are shipped with the program.

What happens in the (rare?) event that the user has edited, say, the supplied water profile that we want to fix?  Do we (can we) detect that, or would their changes by lost by the merge?  (Presumably this is why there's a pop-up warning before we do such merges.)

**Reply by mikfire** [reply]
I do not think it is that rare an occurrence, really. I order hops in bulk and update the parent record every time with the new %AA. I have friends who order grains in bulk, and I believe they do the same.

It looks like we over write any row that we consider ours. Ownership is determined by the bt_* table. You can dig at the code in `Database::updateDatabase` but I think my reading is accurate. 

I do not see anyway we can tell if a specific row has been modified by the user or is just being corrected. All we really can say is the current value is not the new value. I can think of a few ways we could -- like adding meta data on each row to tag if it is unmodified -- but none of that exists now.

I do not find the question we ask explains the risks very well. We just say "There may be new stuff available. Merge?"

**Reply by mikfire** [reply]
So here's my proposal.

I would add a boolean column to the `bt_*` tables called `user_modified` with a default value of `false`. 

If the user edits a namedEntity we check to see if it's id is in the `bt_*` table. if it is, we flip the flag to `true`.

When merging in a database update, we check the `user_modified` in the user's database. If it is set to false, we continue as we do today and just over write it. If the row isn't in the user's database, we continue like we do today and add it.

if the flag is set true, we ask the user if they want to:
- skip the update;
- apply the update, overwriting their changes;
- skip all updates, and stop with the annoying questions;
- apply all updates, and stop with the annoying questions

If they choose one of the 'all' options, we can maybe provide them a list of things we either skipped or applied.  But only those things that had the `user_modified` flag set to true.

I think we may also want to add something to the initial prompt so the user can either say up front they want to skip all conflicts or over write all conflicts.

It does mean there will be one last update that clobbers records, as we have no idea of what the user has modified previously and I would probably populate the new column with `false`.

Having written the code to diff records in the past -- one day, I will resubmit that PR -- I am confident we could try to be more clever and give the user an option as to which attributes they want to preserve or over write on a conflict. I would prefer not to try, as I think it will take some very clever UX design to make it useful. I can do a lot, but clever UX is not on my list of skills.

If we ignore the idea of diff'ing, I can probably get a decent working prototype in a few days and a PR in a week or so.

**Reply by matty0ung** [reply]
The proposal sounds pretty sensible for the problem as stated.  At the same time, I wonder if taking a step back would give us a different problem to solve.  In particular, I had a small and a big question about Hops etc.

One question is how often we make modifications (rather than additions) to the ships-with-the-program data?  If it's only rarely, then maybe it's not worth the effort to be more sophisticated here.  Obviously, this is just me wondering out loud as I don't know what we've had to change in the past.

The other, bigger, question that's been nagging at the back of my mind for a while is whether we ultimately want a different data structure for things like hops.  It seems to me that BeerXML 1.0 over-simplifies the model here, which results in you editing a master hop record every time you buy a new bag of hops.  In a perfect world, might we not split "Hop Variety" out from "Actual Hops"?

Hop Variety would have the description, type, use, origin, substitutes, Hop Storage Index, etc and _ranges_ for typical alpha acid, beta acid, Co-Humulone, Myrcene, etc.  Then Actual Hop (which of course we'd come up with a better name for!) would have harvest year, form (pellets vs leaf), actual alpha acid, actual beta acid, etc.

There might be a similar split for Fermentables, Yeast etc, even if it's smaller amounts of data in Actual Fermentable and Actual Yeast.  Eg, on my original spreadsheet ingredient tracker, I would have "purchase date", "expiry date", "supplier" and "price", and part of my "what to brew next" decision would be based on which ingredients were nearing their expiry date.  But maybe that's just me! :)

Anyway, just throwing it out there as an idea.

**Reply by mikfire** [reply]
First answers first. To my knowledge, we haven't ever shipped a 'correction' to something in the database. We have added a few new things, like the water profiles and some hops. But I do not believe we've actually corrected something. I wouldn't ever trust my memory. You could do a git blame on the default_db.sqlite to see how many times it's been changed.

Oh man. Rethinking the schema is a can of worms, and one I've looked at a few times.

There's a lot of history here. We originally stuck to the beerXML spec because rocketman required it. He made a convincing argument as to why it should be that way. I vaguely remember it had something to do with backwards compatibility with brewtarget 1.x. It's been long enough that I only remember it was a fair argument. He was also the project owner, he said those were the rules and so I either played by them or forked off.

I have made my opinion about beerxml painfully clear (it sucks). I dislike being tied to a poorly written spec that is 20+ years old and no longer reflects modern home brew practices. I think you can see that I have no issues extending it when I like to, like water. My opinion is that we only need to be able to accurately import from and export to beerXML. As long as we translate all the things properly, we can do what ever we like in the database.

Having said that. I am still scarred from the 1.x -> 2.0 transition. It took rocketman and I about a year. We worked on it every night after the day job, and I spent a lot of weekend days on it too. I am unwilling to take on that kind of work without seeing significant improvements to both using brewtarget and maintaining the code. 

Just a thing to consider. Brewing software is trying to solve three different problem domains at the same time. The first problem domain is recipe formulation: build a recipe, doing the math to calculate IBUs, ABV, etc. The second domain is the brew day: heat X amount of water to Y temperature and hold for Z minutes, as well as brew notes. The third domain is inventory: how much of ingredient X do I have on hand, and when does it expire.

What people want from brewtarget depends on how interested they are in each domain. I am deeply interested in the first, somewhat in the second and not at all in the third. That may have been obvious based on how I laid them out :grinning: And just because somebody else is far more interested in the third domain than the first doesn't mean anybody is doing it wrong. It just means I will have a harder time understanding the problems in that domain and then solving them.

Let me poke at your idea for a bit. I am not the sharpest rock in the drawer, and I have found I usually do better for chewing on an idea for a while before reacting.

**Reply by matty0ung** [reply]
I hear what you're saying.  I agree it's not worth doing a big change if it doesn't bring a big benefit.

I know there are some things that deeply interest me as a microbrewer but which don't belong in homebrew software - eg, precise tracking of what's been brewed so I can account for alcohol duty.  (Literally every time I open a bottle, even for myself, I need to make a note so that it can be included on the monthly Customs return.)  One of my long-term projects is to make a downstream project which uses the recipe-formulation and brew-day stuff from Brewtarget but then adds microbrewery-specific functionality (sales, batch tracking for quality control, duty tracking, expiry dates, cost tracking, etc).  Probably going to take a few years, but I hope it will ultimately replace a lot of complicated spreadsheets and Word documents that I use at the moment.

**Reply by mikfire** [reply]
Based on what I am fixing in other places, I am removing the code that updates existing rows in the database. 

Once we ship the data to the user, it isn't ours to touch any more. I would be furious if somebody just blithely over wrote my changes with their data. So we aren't doing that any more. It makes updateDatabase() a bit easier to work through.

**Reply by matty0ung** [reply]
That sounds like a good call.  It feels like, if we have new data to offer to the user, it should be their active choice to import it and, like you say, we shouldn't overwrite any of their existing data.

**Reply by mikfire** [reply]
Actually, we do have one problem.

The BJCP and Brewers Association frequently update their style guides. We will need to figure out some way of updating those. I can see a few ideas, but it will require some thought.

**Comment by matty0ung** [comment]
Another quick question, where does the recipe_children table get used?  (I understand that hop_children, fermentable_children, etc get used when we create a copy of a hop, fermentable, etc to add to a recipe, but since we don't add recipes to recipes, I started wondering about recipe_children.)

**Reply by mikfire** [reply]
I once had a beautiful system for automatic versioning of a recipe. The PR never got merged because I couldn't quite figure out what "delete" meant. 

I also think I spend more time writing and tinkering with recipes than other people, so the idea of an automatic versioning system never really made sense to anybody other than me.

I believe the recipe_children table to be a left over from that effort, and can likely be dropped.

**Comment by matty0ung** [comment]
I think, one way or another, we've now addressed all the issues brought up in this convo, so archiving it.
